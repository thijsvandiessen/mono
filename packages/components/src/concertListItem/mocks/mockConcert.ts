import type { Concert } from '@mono/data'

export const mockConcert: Concert = {
  id: 'id',
  title: 'name',
  image: undefined,
  locations: [
    {
      startTime: '2023-09-01T13:32:36+01:00',
      id: 'test id',
      title: 'test',
      address: 'test',
      ticketLink: 'https://example.com',
      lat: 0,
      lng: 0,
    },
  ],
  url: 'some-slug',
  content: [],
}
