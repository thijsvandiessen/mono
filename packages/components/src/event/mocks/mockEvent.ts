import type { Event } from '@mono/graphql'

export const mockEvent: Event = {
  id: 'id',
  title: 'name',
  image: {
    id: 'image-id',
    title: 'title',
    description: 'desc',
    url: 'https://creaorkest.nl/some-image.png',
  },
  locations: [{ startTime: '2023-09-01T13:32:36+01:00', id: 'test id' }],
  url: 'some-slug',
  content: [],
}
