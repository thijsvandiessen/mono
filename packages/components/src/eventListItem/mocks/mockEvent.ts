import type { Event } from '@crea/graphql/src/types/event'

export const mockEvent: Event = {
  id: 'id',
  title: 'name',
  image: undefined,
  locations: [{ startTime: '2023-09-01T13:32:36+01:00', id: 'test id' }],
  persons: [],
  program: [],
  url: 'some-slug',
  content: [],
}
