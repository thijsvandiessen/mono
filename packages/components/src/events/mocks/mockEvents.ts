import type { Event } from '@crea/graphql/src/types/event'

export const mockEvent: Event = {
  id: 'mock-id',
  title: 'title',
  url: '/concerten/concert-1',
  image: undefined,
  locations: [],
  persons: [],
  program: [],
}
