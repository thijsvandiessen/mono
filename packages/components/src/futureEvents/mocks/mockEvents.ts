import { type Event } from '@crea/graphql/src/types/event'

export const mockEvent: Event = {
  id: 'mock-id',
  title: 'title',
  image: undefined,
  locations: [],
  persons: [],
  program: [],
  url: '/some-slug',
  content: [],
}
