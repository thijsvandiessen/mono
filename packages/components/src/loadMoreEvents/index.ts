export { LoadMoreEvents } from './loadMoreEvents.jsx'
