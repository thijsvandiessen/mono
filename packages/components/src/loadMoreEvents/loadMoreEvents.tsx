'use client'
import { EventListItem } from '../eventListItem'
import type { Event as EventType } from '@mono/graphql'
import React, { useEffect, useRef, useState } from 'react'
import { getEvents } from '@mono/graphql'
import { useEventsMeta, useIntersectionObserver } from '@mono/hooks'
import { ConcertModelOrderBy } from '@mono/graphql'

export interface Props {
  initialSkip: number
}

export const LoadMoreEvents = ({ initialSkip }: Props) => {
  const [skip, setSkip] = useState(initialSkip)
  const [loading, setLoading] = useState(false)
  const [events, setEvents] = useState<(EventType | undefined)[]>([])
  const ref = useRef<HTMLDivElement | null>(null)
  const { numberOfEvents } = useEventsMeta()
  const [entry] = useIntersectionObserver({
    enabled: Boolean(numberOfEvents),
    ref,
  })

  useEffect(() => {
    const interval = 10
    if (loading) return
    if (!entry?.isIntersecting) return
    if (!numberOfEvents) return
    if (events.length === numberOfEvents - initialSkip) return
    if (skip > numberOfEvents) return
    setLoading(true)
    // TODO: abort signal to stop fetching
    getEvents({
      skip,
      first: interval,
      order: [ConcertModelOrderBy.PositionAsc],
    })
      .then(({ data }) => {
        if (!data) return
        setEvents((prev) => [...prev, ...data])
        setSkip((prev) => {
          return prev + interval
        })
      })
      .catch(() => {
        console.log('TODO: render an error')
      })
      .finally(() => {
        setLoading(false)
      })
  }, [
    loading,
    skip,
    numberOfEvents,
    entry?.isIntersecting,
    events.length,
    initialSkip,
  ])

  return (
    <>
      {events.map((event, index) => {
        if (!event?.id) return
        return (
          <EventListItem
            className="content-layout--small"
            key={event.id}
            data={event}
            isLast={events.length - 1 === index}
          />
        )
      })}

      {numberOfEvents ? (
        <div ref={ref}>
          <p className="text-small">
            {/* TODO: simplify/fix this logic */}
            {events.length + initialSkip} van {numberOfEvents} concerten
          </p>
        </div>
      ) : (
        <p className="text-small">...</p>
      )}
    </>
  )
}
