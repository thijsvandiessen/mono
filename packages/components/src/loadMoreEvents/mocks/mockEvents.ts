import { type Event } from '@crea/graphql/src/types/event'

export const mockEvent: Event = {
  id: 'mock-id',
  title: 'title',
  url: '/url',
  image: undefined,
  locations: [],
  persons: [],
  program: [],
}
