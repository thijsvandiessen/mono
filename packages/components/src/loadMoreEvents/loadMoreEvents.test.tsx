import { LoadMoreEvents } from './loadMoreEvents'
import React from 'react'
import { getEvents } from '@mono/graphql'
import { mockEvent } from './mocks/mockEvents'
import { render, screen } from '@testing-library/react'
import { describe, it, expect, vi } from 'vitest'

vi.mock('@mono/graphql', () => ({
  __esModule: true,
  ...vi.importActual('@mono/graphql'),
  getEvents: vi.fn(),
}))

vi.mock('@mono/hooks', () => ({
  __esModule: true,
  ...vi.importActual('@mono/hooks'),
  useEventsMeta: vi.fn(() => ({ numberOfEvents: 10 })),
  useIntersectionObserver: vi.fn(() => []),
}))

const getEventsMock = vi.mocked(getEvents)

describe('Events component', () => {
  it('shows all the data', () => {
    getEventsMock.mockResolvedValue({
      data: [mockEvent],
      error: undefined,
    })

    const { container } = render(<LoadMoreEvents initialSkip={1} />)

    expect(screen.getByText('1 van 10 concerten')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
