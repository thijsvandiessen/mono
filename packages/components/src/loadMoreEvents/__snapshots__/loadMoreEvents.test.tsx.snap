// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events component shows all the data 1`] = `
<div>
  <div>
    <p
      class="text-small"
    >
      1
       van 
      10
       concerten
    </p>
  </div>
</div>
`;
